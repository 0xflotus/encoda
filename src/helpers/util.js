const path = require('path')

function replaceExt (path_, ext) {
  let match = path.basename(path_).match(/^([^.]+)?\.(.+)$/)
  if (match) return path.join(path.dirname(path_), match[1] + ext)
  else return path_ + ext
}

module.exports = {
  replaceExt
}
