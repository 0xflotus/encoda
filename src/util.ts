import * as stencila from '@stencila/schema'
import Ajv from 'ajv'
import betterAjvErrors from 'better-ajv-errors'
import { record } from 'fp-ts'
import { eqString } from 'fp-ts/lib/Eq'
import { pipe } from 'fp-ts/lib/pipeable'
import fs from 'fs-extra'
import globby from 'globby'
import produce from 'immer'
import path from 'path'
import { decode as decodePerson } from './person'
import type from './util/type'

const built = path.join(
  path.dirname(require.resolve('@stencila/schema')),
  'built'
)

// Load all schemas for use in by Ajv validator
const schemas = globby
  .sync(path.join(built, '*.schema.json'))
  .map(file => fs.readJSONSync(file))

// Read in aliases for use in coerce function
const aliases = fs.readJSONSync(path.join(built, 'aliases.json'))

/**
 * Create a node of a type
 * @param type The name of the type
 * @param initial Initial values for properties
 * @param validation What validation should done?
 */
export function create<Key extends keyof stencila.Types>(
  type: Key,
  initial: { [key: string]: any } = {},
  validation: 'none' | 'validate' | 'coerce' = 'validate'
): stencila.Types[Key] {
  let node = { type, ...initial }
  if (validation === 'validate') return validate(node, type)
  else if (validation === 'coerce') return coerce(node, type)
  else return node
}

/**
 * Cast a node to a particular type
 *
 * The node is validated against the type.
 * This means that an error will be throw if during:
 *   - up-casting the node does not have properties
 *     that are required by the schema of the new type
 *   - down-casting the node has properties that
 *     are additional to those in the schema of the new type
 * Use `coerce` if you want to ignore such errors
 * and force mutating the node to the type.
 *
 * @param node The node to cast
 * @param type The type to cast to
 */
export function cast<Key extends keyof stencila.Types>(
  node: any,
  type: Key
): stencila.Types[Key] {
  return produce(node, (casted: any) => {
    casted.type = type
    validate(casted, type)
  })
}

// Cached JSON Schema validation functions
const validators = new Ajv({
  schemas,
  jsonPointers: true
})

/**
 * Validate a node against a type's schema
 * @param node The node to validate
 * @param type The type to validate against
 */
export function validate<Key extends keyof stencila.Types>(
  node: any,
  type: Key
): stencila.Types[Key] {
  const validator = validators.getSchema(
    `https://stencila.github.com/schema/${type}.schema.json`
  )
  if (!validator) throw new Error(`No schema for type "${type}".`)
  if (!validator(node)) {
    const errors = (betterAjvErrors(validator.schema, node, validator.errors, {
      format: 'js'
    }) as unknown) as betterAjvErrors.IOutputError[]
    throw new Error(errors.map(error => `${error.error}`).join(';'))
  }
  return node
}

/**
 * Is a node valid with respect to a particular type's schema
 * @param node The node to check
 * @param type The type to check against
 */
export function valid<Key extends keyof stencila.Types>(
  node: any,
  type: Key
): boolean {
  try {
    validate(node, type)
    return true
  } catch (error) {
    return false
  }
}

// Cached JSON Schema validation/mutation functions
// These use Ajv options that coerce nodes so we
// keep them separate from pure non-mutating validators.
const mutators = new Ajv({
  schemas,
  jsonPointers: true,
  // Add values from `default` keyword when property is missing
  useDefaults: true,
  // Remove any additional properties
  removeAdditional: true,
  // Coerce type of data to match type keyword and coerce scalar
  // data to an array with one element and vice versa, as needed.
  coerceTypes: 'array'
})

/**
 * A list of decoders that can be applied using the `decoder` keyword
 */
const decoders: { [key: string]: (data: string) => any } = {
  csv,
  ssv,
  person: decodePerson
}

/**
 * Decode comma separated string data into an array of strings
 */
function csv(data: string): Array<string> {
  return data.split(',')
}

/**
 * Decode space separated string data into an array of strings
 */
function ssv(data: string): Array<string> {
  return data.split(/ +/)
}

/**
 * Custom validation function that handles the `decoder`
 * keyword.
 */
const decoderValidate: Ajv.SchemaValidateFunction = (
  decoder: string,
  data: string,
  parentSchema?: object,
  dataPath?: string,
  parentData?: object | Array<any>,
  parentDataProperty?: string | number,
  rootData?: object | Array<any>
): boolean => {
  function raise(msg: string) {
    decoderValidate.errors = [
      {
        keyword: 'parser',
        dataPath: '' + dataPath,
        schemaPath: '',
        params: {
          keyword: 'parser'
        },
        message: msg,
        data: data
      }
    ]
    return false
  }
  const decode = decoders[decoder]
  if (!decode) return raise(`no such decoder: "${decoder}"`)

  let decoded: any
  try {
    decoded = decode(data)
  } catch (error) {
    const decodeError = error.message.split('\n')[0]
    return raise(`error when decoding using "${decoder}": ${decodeError}`)
  }

  if (parentData !== undefined && parentDataProperty !== undefined) {
    ;(parentData as any)[parentDataProperty] = decoded
  }
  return true
}

mutators.addKeyword('parser', {
  type: 'string',
  modifying: true,
  validate: decoderValidate
})

/**
 * Coerce a node so it conforms to a type's schema
 *
 * @param node The node to coerce
 * @param typeName The type to coerce it to
 */
export function coerce<Key extends keyof stencila.Types>(
  node: any,
  typeName?: Key
): stencila.Types[Key] {
  if (!typeName) typeName = type(node) as Key

  const mutator = mutators.getSchema(
    `https://stencila.github.com/schema/${typeName}.schema.json`
  )
  if (!mutator) throw new Error(`No schema for type "${typeName}".`)

  return produce(node, (coerced: any) => {
    if (typeof coerced === 'object') coerced.type = typeName
    // Rename property aliases
    rename(coerced)
    // coerce and validate
    if (!mutator(coerced)) {
      const errors = (betterAjvErrors(mutator.schema, node, mutator.errors, {
        format: 'js'
      }) as unknown) as betterAjvErrors.IOutputError[]
      throw new Error(errors.map(error => `${error.error}`).join(';'))
    }
  })
  // Replace aliases with canonical names
  function rename(node: any) {
    if (!node || typeof node !== 'object') return
    if (!(node.type && aliases[node.type])) return

    let propertyAliases = aliases[node.type]
    for (let [key, child] of Object.entries(node)) {
      if (!Array.isArray(node)) {
        const name = propertyAliases[key]
        if (name) {
          node[name] = child
          delete node[key]
        }
      }
      rename(child)
    }
  }
}

/* Wrap non-`BlockContent` nodes in `Paragraph` nodes to conform to schema */
export const wrapInBlockNode = (node: stencila.Node): stencila.BlockContent => {
  return isBlockContent(node)
    ? node
    : { type: 'Paragraph', content: [node].filter(isInlineContent) }
}

export const hasType = (node?: stencila.Node): node is stencila.Thing => {
  if (!node) return false
  if (Array.isArray(node)) return false
  if (typeof node !== 'object') return false
  if (!node.type) return false
  return true
}

export const isNodeType = <Ts extends { type: string }>(
  typeMap: { [key in Ts['type']]: key }
) => (nodeType: string): boolean => {
  return pipe(record.elem(eqString)(nodeType, typeMap))
}

type NodeType = { type: string } & { [key: string]: unknown }

const hasTypeProp = (o: { type?: string }): o is NodeType =>
  o.type ? true : false

export const isNode = <T extends object, Ts extends NodeType = NodeType>(
  typeMap: { [key in Ts['type']]: key }
) => (n: unknown): n is T => {
  if (!n) return false
  if (Array.isArray(n)) return false
  if (typeof n !== 'object') return false
  if (n === null) return false
  return !hasTypeProp(n) ? false : isNodeType<Ts>(typeMap)(n.type)
}

export const blockContentTypes: {
  [key in stencila.BlockContent['type']]: key
} = {
  CodeBlock: 'CodeBlock',
  CodeChunk: 'CodeChunk',
  Heading: 'Heading',
  List: 'List',
  ListItem: 'ListItem',
  Paragraph: 'Paragraph',
  QuoteBlock: 'QuoteBlock',
  Table: 'Table',
  ThematicBreak: 'ThematicBreak'
}

export const isBlockContent = isNode<stencila.BlockContent>(blockContentTypes)

type InlineNodesWithType = Exclude<
  stencila.InlineContent,
  string | null | boolean | number
>['type']

export const inlineContentTypes: { [key in InlineNodesWithType]: key } = {
  Code: 'Code',
  CodeBlock: 'CodeBlock',
  CodeExpr: 'CodeExpr',
  Delete: 'Delete',
  Emphasis: 'Emphasis',
  ImageObject: 'ImageObject',
  Link: 'Link',
  Quote: 'Quote',
  Strong: 'Strong'
}

// null | boolean | string | number
export const isInlinePrimitive = (
  node: stencila.Node
): node is null | boolean | number | string => {
  if (node === undefined) return false
  if (node === null) return true
  if (typeof node === 'boolean') return true
  if (typeof node === 'number') return true
  if (typeof node === 'string') return true
  return false
}

export const isInlineNonPrimitive = (
  node: stencila.Node
): node is InlineNodesWithType => {
  if (isInlinePrimitive(node)) return false
  if (Array.isArray(node)) return false
  if (typeof node === 'object' && hasTypeProp(node))
    return isNodeType(inlineContentTypes)(node.type)
  return false
}

export const isInlineContent = (
  node: stencila.Node
): node is stencila.InlineContent => {
  return isInlinePrimitive(node) || isInlineNonPrimitive(node)
}
