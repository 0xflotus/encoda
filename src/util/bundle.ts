import * as stencila from '@stencila/schema'
import produce from 'immer'
import * as dataUri from './dataUri'

/**
 * Walk a document tree and replace any links to local resources
 * with data URIs.
 *
 * This is used to create standalone HTML pages
 * as well as when using Puppeteer which does not
 * load local resources for security reasons.
 * See https://github.com/GoogleChrome/puppeteer/issues/1643.
 *
 * Currently only handles `MediaObject` nodes, but could be used for other
 * node types in the future.
 */
export default async function bundle(
  node: stencila.Node
): Promise<stencila.Node> {
  async function walk(node: any) {
    if (node === null || typeof node !== 'object') return node

    switch (node.type) {
      case 'MediaObject':
      case 'AudioObject':
      case 'ImageObject':
      case 'VideoObject':
        const mediaObject = node as stencila.MediaObject
        if (!mediaObject.contentUrl.startsWith('http')) {
          const { dataUri: contentUrl } = await dataUri.fromFile(
            mediaObject.contentUrl
          )
          return {
            ...mediaObject,
            contentUrl
          }
        }
    }

    for (const [key, child] of Object.entries(node)) {
      node[key] = await walk(child)
    }
    return node
  }
  return await produce(node, walk)
}
