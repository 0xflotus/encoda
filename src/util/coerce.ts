import stencila, { isEntity, nodeType } from '@stencila/schema'
import Ajv from 'ajv'
import produce from 'immer'
import { getCoecer, getErrors, getSchema } from './schemas'

/**
 * Coerce a node so it conforms to a type's schema
 *
 * @param node The node to coerce
 * @param typeName The type to coerce it to
 */
export async function coerce<Key extends keyof stencila.Types>(
  node: stencila.Node,
  type?: Key
): Promise<stencila.Types[Key]> {
  if (type === undefined) type = nodeType(node) as Key
  const coecer = await getCoecer(type)

  return (produce(node, async (coerced: stencila.Node) => {
    if (coerced === null) return node
    // Change the type, since this is coercion to the specified type
    // @ts-ignore
    if (isEntity(coerced)) coerced.type = type
    // Rename property aliases
    await rename(coerced)
    // Coerce and validate
    try {
      return await coecer(coerced)
    } catch (error) {
      if (error instanceof Ajv.ValidationError)
        throw getErrors(coecer, node, error.errors)
      else throw error
    }
  }) as unknown) as stencila.Types[Key]

  // Replace aliases with canonical names
  async function rename(node: stencila.Node): Promise<void> {
    if (isEntity(node)) {
      const schema = await getSchema(node.type as Key)
      if (!schema.propertyAliases) return

      for (const [key, child] of Object.entries(node)) {
        const name = schema.propertyAliases[key]
        if (name) {
          // @ts-ignore
          node[name] = child
          // @ts-ignore
          delete node[key]
        }
        await rename(child)
      }
    } else if (Array.isArray(node)) {
      for (const child of node) {
        await rename(child)
      }
    }
  }
}
