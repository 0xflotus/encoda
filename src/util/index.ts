/**
 * @module util
 */

// TODO: This file uses any a lot. Needs a refactor and re-enable linting
/* eslint-disable
      @typescript-eslint/no-explicit-any,
      @typescript-eslint/explicit-function-return-type,
      @typescript-eslint/no-use-before-define
*/

import * as stencila from '@stencila/schema'
import Ajv from 'ajv'
import betterAjvErrors from 'better-ajv-errors'
import { record } from 'fp-ts'
import { eqString } from 'fp-ts/lib/Eq'
import { pipe } from 'fp-ts/lib/pipeable'
import fs from 'fs-extra'
import produce from 'immer'
import path from 'path'
import { decode as decodePerson } from '../codecs/person'
import type from './type'

/**
 * Create a node of a type
 * @param type The name of the type
 * @param initial Initial values for properties
 * @param validation What validation should done?
 */
export async function create<Key extends keyof stencila.Types>(
  type: Key,
  initial: { [key: string]: any } = {},
  validation: 'none' | 'validate' | 'coerce' = 'validate'
): Promise<stencila.Types[Key]> {
  let node = { type, ...initial }
  if (validation === 'validate') return validate(node, type)
  else if (validation === 'coerce') return coerce(node, type)
  else return node
}

/**
 * Cast a node to a particular type
 *
 * The node is validated against the type.
 * This means that an error will be throw if during:
 *   - up-casting the node does not have properties
 *     that are required by the schema of the new type
 *   - down-casting the node has properties that
 *     are additional to those in the schema of the new type
 * Use `coerce` if you want to ignore such errors
 * and force mutating the node to the type.
 *
 * @param node The node to cast
 * @param type The type to cast to
 */
export async function cast<Key extends keyof stencila.Types>(
  node: any,
  type: Key
): Promise<stencila.Types[Key]> {
  return produce(node, async (casted: any) => {
    casted.type = type
    await validate(casted, type)
  })
}

// Cached JSON Schema validation functions
const validators = new Ajv({
  jsonPointers: true,
  loadSchema
})

/**
 * Load a JSON Schema based on its URI
 */
async function loadSchema(uri: string) {
  const match = uri.match(/([\w]+)\.schema\.json$/)
  if (match) return readSchema(match[1])
  throw new Error(`Can not resolve schema "${uri}"`)
}

const schemasPath = path.dirname(require.resolve('@stencila/schema'))

/**
 * Read a JSON Schema file from `@stencila/schema`
 */
async function readSchema<Key extends keyof stencila.Types>(type: Key) {
  try {
    return await fs.readJSON(path.join(schemasPath, `${type}.schema.json`))
  } catch (error) {
    if (error.code === 'ENOENT')
      throw new Error(`No schema for type "${type}".`)
    throw error
  }
}

/**
 * Get the `Ajv` validation function for a type
 */
async function getValidator<Key extends keyof stencila.Types>(
  ajv: Ajv.Ajv,
  type: Key
): Promise<Ajv.ValidateFunction> {
  let validator = ajv.getSchema(
    `https://stencila.github.com/schema/${type}.schema.json`
  )
  if (!validator) {
    const schema = await readSchema(type)
    validator = await ajv.compileAsync(schema)
  }
  return validator
}

/**
 * Get the `Schema` object for a type
 */
async function getSchema<Key extends keyof stencila.Types>(
  ajv: Ajv.Ajv,
  type: Key
): Promise<any> {
  const validator = await getValidator(ajv, type)
  const schema = validator.schema
  if (schema === undefined || typeof schema === 'boolean')
    throw new Error(`Woaah! No schema on validator for type "${type}".`)
  return schema
}

/**
 * Validate a node against a type's schema
 * @param node The node to validate
 * @param type The type to validate against
 */
export async function validate<Key extends keyof stencila.Types>(
  node: any,
  typeName?: Key
): Promise<stencila.Types[Key]> {
  if (typeName === undefined) typeName = type(node) as Key
  const validator = await getValidator(validators, typeName)
  if (!validator(node)) {
    const errors = (betterAjvErrors(validator.schema, node, validator.errors, {
      format: 'js'
    }) as unknown) as betterAjvErrors.IOutputError[]
    throw new Error(errors.map(error => `${error.error}`).join(';'))
  }
  return node
}

/**
 * Is a node valid with respect to a particular type's schema
 * @param node The node to check
 * @param type The type to check against
 */
export async function valid<Key extends keyof stencila.Types>(
  node: any,
  type: Key
): Promise<boolean> {
  try {
    await validate(node, type)
    return true
  } catch (error) {
    return false
  }
}

// Cached JSON Schema validation/mutation functions
// These use Ajv options that coerce nodes so we
// keep them separate from pure non-mutating validators.
const mutators = new Ajv({
  jsonPointers: true,
  // Add values from `default` keyword when property is missing
  useDefaults: true,
  // Remove any additional properties
  removeAdditional: true,
  // Coerce type of data to match type keyword and coerce scalar
  // data to an array with one element and vice versa, as needed.
  coerceTypes: 'array',
  loadSchema
})

/**
 * A list of codecs that can be applied using the `codec` keyword
 *
 * TODO: Make these actual codecs (i.e. conforming to the `Codec`
 * interface with a `decode` async function)
 */
const codecs: { [key: string]: (data: string) => any } = {
  csv,
  ssv,
  person: decodePerson
}

/**
 * Decode comma separated string data into an array of strings
 */
function csv(data: string): string[] {
  return data.split(',')
}

/**
 * Decode space separated string data into an array of strings
 */
function ssv(data: string): string[] {
  return data.split(/ +/)
}

/**
 * Custom validation function that handles the `codec`
 * keyword.
 */
const codecValidate: Ajv.SchemaValidateFunction = (
  codec: string,
  data: string,
  parentSchema?: object,
  dataPath?: string,
  parentData?: object | any[],
  parentDataProperty?: string | number
  // rootData?: object | any[]
): boolean => {
  function raise(msg: string) {
    codecValidate.errors = [
      {
        keyword: 'decoding',
        dataPath: dataPath || '',
        schemaPath: '',
        params: {
          keyword: 'codec'
        },
        message: msg,
        data: data
      }
    ]
    return false
  }
  const decode = codecs[codec]
  if (!decode) return raise(`no such codec: "${codec}"`)

  let decoded: any
  try {
    decoded = decode(data)
  } catch (error) {
    const decodeError = error.message.split('\n')[0]
    return raise(`error using "${codec}" codec: ${decodeError}`)
  }

  if (parentData !== undefined && parentDataProperty !== undefined) {
    ;(parentData as any)[parentDataProperty] = decoded
  }
  return true
}

mutators.addKeyword('codec', {
  type: 'string',
  modifying: true,
  validate: codecValidate
})

/**
 * Coerce a node so it conforms to a type's schema
 *
 * @param node The node to coerce
 * @param typeName The type to coerce it to
 */
export async function coerce<Key extends keyof stencila.Types>(
  node: any,
  typeName?: Key
): Promise<stencila.Types[Key]> {
  if (typeName === undefined) typeName = type(node) as Key
  const mutator = await getValidator(mutators, typeName)

  return produce(node, async (coerced: any) => {
    if (typeof coerced === 'object') coerced.type = typeName
    // Rename property aliases
    await rename(coerced)
    // coerce and validate
    if (!mutator(coerced)) {
      const errors = (betterAjvErrors(mutator.schema, node, mutator.errors, {
        format: 'js'
      }) as unknown) as betterAjvErrors.IOutputError[]
      throw new Error(errors.map(error => `${error.error}`).join(';'))
    }
  })
  // Replace aliases with canonical names
  async function rename(node: any) {
    if (!node || typeof node !== 'object') return
    if (!node.type) return
    const schema = await getSchema(mutators, node.type as Key)
    if (!schema.propertyAliases) return

    for (let [key, child] of Object.entries(node)) {
      if (!Array.isArray(node)) {
        const name = schema.propertyAliases[key]
        if (name) {
          node[name] = child
          delete node[key]
        }
      }
      await rename(child)
    }
  }
}

/* Wrap non-`BlockContent` nodes in `Paragraph` nodes to conform to schema */
export const wrapInBlockNode = (node: stencila.Node): stencila.BlockContent => {
  return isBlockContent(node)
    ? node
    : { type: 'Paragraph', content: [node].filter(isInlineContent) }
}

export const hasType = (node?: stencila.Node): node is stencila.Thing => {
  if (!node) return false
  if (Array.isArray(node)) return false
  if (typeof node !== 'object') return false
  if (!node.type) return false
  return true
}

export const isNodeType = <Ts extends { type: string }>(
  typeMap: { [key in Ts['type']]: key }
) => (nodeType: string): boolean => {
  return pipe(record.elem(eqString)(nodeType, typeMap))
}

type NodeType = { type: string } & { [key: string]: unknown }

const hasTypeProp = (o: { type?: string }): o is NodeType => !!o.type

export const isNode = <T extends object, Ts extends NodeType = NodeType>(
  typeMap: { [key in Ts['type']]: key }
) => (n: unknown): n is T => {
  if (!n) return false
  if (Array.isArray(n)) return false
  if (typeof n !== 'object') return false
  if (n === null) return false
  return !hasTypeProp(n) ? false : isNodeType<Ts>(typeMap)(n.type)
}

export const blockContentTypes: {
  [key in stencila.BlockContent['type']]: key
} = {
  CodeBlock: 'CodeBlock',
  CodeChunk: 'CodeChunk',
  Heading: 'Heading',
  List: 'List',
  ListItem: 'ListItem',
  Paragraph: 'Paragraph',
  QuoteBlock: 'QuoteBlock',
  Table: 'Table',
  ThematicBreak: 'ThematicBreak'
}

export const isBlockContent = isNode<stencila.BlockContent>(blockContentTypes)

type InlineNodesWithType = Exclude<
  stencila.InlineContent,
  string | null | boolean | number
>['type']

export const inlineContentTypes: { [key in InlineNodesWithType]: key } = {
  Code: 'Code',
  CodeBlock: 'CodeBlock',
  CodeExpr: 'CodeExpr',
  Delete: 'Delete',
  Emphasis: 'Emphasis',
  ImageObject: 'ImageObject',
  Link: 'Link',
  Quote: 'Quote',
  Strong: 'Strong',
  Subscript: 'Subscript',
  Superscript: 'Superscript'
}

// null | boolean | string | number
export const isInlinePrimitive = (
  node: stencila.Node
): node is null | boolean | number | string => {
  if (node === undefined) return false
  if (node === null) return true
  if (typeof node === 'boolean') return true
  if (typeof node === 'number') return true
  if (typeof node === 'string') return true
  return false
}

export const isInlineNonPrimitive = (
  node: stencila.Node
): node is InlineNodesWithType => {
  if (isInlinePrimitive(node)) return false
  if (Array.isArray(node)) return false
  if (typeof node === 'object' && hasTypeProp(node))
    return isNodeType(inlineContentTypes)(node.type)
  return false
}

export const isInlineContent = (
  node: stencila.Node
): node is stencila.InlineContent => {
  return isInlinePrimitive(node) || isInlineNonPrimitive(node)
}

export const creativeWorkTypes: {
  [key in stencila.CreativeWork['type']]: key
} = {
  CreativeWork: 'CreativeWork',
  Article: 'Article',
  AudioObject: 'AudioObject',
  CodeChunk: 'CodeChunk',
  CodeExpr: 'CodeExpr',
  Collection: 'Collection',
  Datatable: 'Datatable',
  ImageObject: 'ImageObject',
  MediaObject: 'MediaObject',
  SoftwareApplication: 'SoftwareApplication',
  SoftwareSourceCode: 'SoftwareSourceCode',
  Table: 'Table',
  VideoObject: 'VideoObject'
}

export const isCreativeWork = isNode<stencila.CreativeWork>(creativeWorkTypes)
