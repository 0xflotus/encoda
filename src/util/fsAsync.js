import fs from 'fs'
import {promisify} from 'util'

fs.readFileAsync = promisify(fs.readFile)
fs.writeFileAsync = promisify(fs.writeFile)

export default fs
