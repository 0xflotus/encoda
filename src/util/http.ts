/**
 * A utility module for HTTP requests
 *
 * Uses `got` to enable RFC 7234 compliant HTTP caching
 *
 * @module util/http
 */

import fs from 'fs-extra'
import got from 'got'
import stream from 'stream'
import util from 'util'
import cache from './app/cache'
import packg from '../../package.json'

const pipeline = util.promisify(stream.pipeline)

const http = got.extend({
  cache,
  headers: {
    'user-agent': `encoda/${packg.version} (https://github.com/stencila/encoda)`,
    'accept-encoding': 'gzip, deflate'
  }
})

/**
 * Get content from a URL
 *
 * @param url The URL to get
 * @param options Options to pass to `got`
 */
export async function get(url: string, options = {}) {
  return await http.get(url, options)
}

/**
 * Download a file
 *
 * @param url The URL to download from
 * @param filePath The file path to download to
 */
export async function download(url: string, filePath: string) {
  return pipeline(http.stream(url), fs.createWriteStream(filePath))
}

/**
 * Clear any cached content for a particular URL
 *
 * @param url The URL to clear the cache for
 */
export async function cacheDelete(url: string) {
  cache.delete('cacheable-request:GET:' + url)
}
