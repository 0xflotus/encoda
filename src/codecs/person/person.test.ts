import { decode } from './'
import { coerce, create, validate } from '../../util/index'

describe('validate', () => {
  it('throws for invalid emails', async () => {
    await expect(
      validate(
        {
          type: 'Person',
          emails: ['pete@example.com', 'pete_at_example_com']
        },
        'Person'
      )
    ).rejects.toThrow('/emails/1: format should match format "email"')
  })
})

describe('decode', () => {
  it('works', async () => {
    let person = await create('Person')

    person.familyNames = ['Jones']
    expect(decode('Jones')).toEqual(person)

    person.givenNames = ['Jane', 'Jill']
    expect(decode('Jane Jill Jones')).toEqual(person)

    person.honorificPrefix = 'Dr'
    expect(decode('Dr Jane Jill Jones')).toEqual(person)

    person.honorificSuffix = 'PhD'
    expect(decode('Dr Jane Jill Jones PhD')).toEqual(person)

    person.emails = ['jane@example.com']
    expect(decode('Dr Jane Jill Jones PhD <jane@example.com>')).toEqual(person)

    person.url = 'http://example.com/jane'
    expect(
      decode(
        'Dr Jane Jill Jones PhD <jane@example.com> (http://example.com/jane)'
      )
    ).toEqual(person)
  })

  it('throws', () => {
    expect(() => decode('')).toThrow(
      /^Unable to decode string \"\" as a person$/
    )
    expect(() => decode('#@&%')).toThrow(
      /^Unable to decode string \"#@&%\" as a person$/s
    )
  })
})

describe('coerce', () => {
  it('coerces properties', async () => {
    expect(
      await coerce(
        {
          givenNames: 'John Tom',
          familyNames: 'Smith',
          // Unfortunately it is not possible to
          // coerce an an object to an array of objects
          // so we must always have an array here
          affiliations: [
            {
              name: 'University of Beep, Boop'
            }
          ]
        },
        'Person'
      )
    ).toEqual({
      type: 'Person',
      givenNames: ['John', 'Tom'],
      familyNames: ['Smith'],
      affiliations: [
        {
          type: 'Organization',
          name: 'University of Beep, Boop'
        }
      ]
    })
  })

  it('renames and coerces property aliases', async () => {
    expect(
      await coerce(
        {
          firstNames: 'John Tom',
          lastName: 'Smith'
        },
        'Person'
      )
    ).toEqual({
      type: 'Person',
      givenNames: ['John', 'Tom'],
      familyNames: ['Smith']
    })

    expect(
      await coerce(
        {
          givenName: 'Jane',
          surnames: 'Doe Smith'
        },
        'Person'
      )
    ).toEqual({
      type: 'Person',
      givenNames: ['Jane'],
      familyNames: ['Doe', 'Smith']
    })
  })

  it('parses strings into people', async () => {
    expect(
      await coerce(
        {
          authors: [
            'John Smith',
            'Dr Jane Jones PhD <jane@example.com>',
            'Jones, Jack (http://example.com/jack)'
          ]
        },
        'CreativeWork'
      )
    ).toEqual({
      type: 'CreativeWork',
      authors: [
        {
          type: 'Person',
          givenNames: ['John'],
          familyNames: ['Smith']
        },
        {
          type: 'Person',
          honorificPrefix: 'Dr',
          givenNames: ['Jane'],
          familyNames: ['Jones'],
          honorificSuffix: 'PhD',
          emails: ['jane@example.com']
        },
        {
          type: 'Person',
          givenNames: ['Jack'],
          familyNames: ['Jones'],
          url: 'http://example.com/jack'
        }
      ]
    })
  })

  it('throws if string can not be parsed', async () => {
    await expect(
      coerce({ authors: ['John Smith', '#@&%', 'Jones, Jane'] }, 'CreativeWork')
    ).rejects.toThrow(
      '/authors/1: parser error when decoding using "person": Unable to decode string "#@&%" as a person'
    )
  })
})
