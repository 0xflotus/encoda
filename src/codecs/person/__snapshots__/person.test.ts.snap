// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decode decodes an orcid to a Person 1`] = `
Object {
  "affiliations": Array [
    Object {
      "alternateNames": Array [
        "Psychoceramics",
      ],
      "identifiers": Array [
        Object {
          "propertyID": "RINGGOLD",
          "type": "PropertyValue",
          "value": 5468,
        },
      ],
      "name": "Wesleyan University",
      "type": "Organization",
    },
    Object {
      "alternateNames": Array [
        "Psychoceramics",
      ],
      "identifiers": Array [
        Object {
          "propertyID": "RINGGOLD",
          "type": "PropertyValue",
          "value": 6752,
        },
      ],
      "name": "Brown University",
      "type": "Organization",
    },
  ],
  "alternateNames": Array [
    "Josiah Stinkney Carberry",
    "J. Carberry",
    "J. S. Carberry",
  ],
  "familyNames": Array [
    "Carberry",
  ],
  "givenNames": Array [
    "Josiah",
  ],
  "id": "https://orcid.org/0000-0002-1825-0097",
  "identifiers": Array [
    Object {
      "propertyID": "Scopus Author ID",
      "type": "PropertyValue",
      "value": 7007156898,
    },
  ],
  "type": "Person",
  "url": "http://library.brown.edu/about/hay/carberry.php",
}
`;
